plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php

```diff
@@ -0,0 +1,317 @@
+	 *
+	 * The returned array will contain the following (only 'install_ok' and 'messages' if the installation fails):
+	 *
+	 * - 'install_ok', a boolean.
+	 * - 'messages', all the messages generated by the installer.
+	 * - 'plugin_name', useful if it wasn't supplied in $metadata.
+	 * - 'version', of the plugin that has been installed.
+	 * - 'date', ISO-formatted installation date.
+	 * - 'metadata', as supplied (except the 'plugin_name' key) and only if not empty.
+	 *
+	 * @param string $plugin_url URL or file path of the plugin to install.
+	 * @param array  $metadata Metadata to store if the installation succeeds.
+	 * @return array Information about the installation result.
+	 */
+	public function install_plugin( string $plugin_url, array $metadata = array() ): array {
+		$this->installing_plugin = true;
+		try {
+			return $this->install_plugin_core( $plugin_url, $metadata );
+		} finally {
+			$this->installing_plugin = false;
+		}
+	}
+
+	/**
+	 * Core version of 'install_plugin' (it doesn't handle the $installing_plugin flag).
+	 *
+	 * @param string $plugin_url URL or file path of the plugin to install.
+	 * @param array  $metadata Metadata to store if the installation succeeds.
+	 * @return array Information about the installation result.
+	 */
+	private function install_plugin_core( string $plugin_url, array $metadata ): array {
+		wp_clean_plugins_cache();
+
+		$plugin_name = $metadata['plugin_name'] ?? null;
+		unset( $metadata['plugin_name'] );
+		if ( is_null( $plugin_name ) ) {
+			$old_plugin_names = array_keys( get_plugins() );
+		} elseif ( ! is_null( $this->get_installed_plugin_info( $plugin_name ) ) ) {
+			return array(
+				'install_ok'         => false,
+				'installer_messages' => __( 'Plugin is already installed', 'woocommerce' ),
+			);
+		}
+
+		include_once ABSPATH . 'wp-admin/includes/file.php';
+		include_once ABSPATH . 'wp-admin/includes/misc.php';
+		include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+		include_once ABSPATH . 'wp-admin/includes/class-automatic-upgrader-skin.php';
+		$skin     = new \Automatic_Upgrader_Skin();
+		$upgrader = new \Plugin_Upgrader( $skin );
+
+		$install_ok = $upgrader->install( $plugin_url );
+		wp_clean_plugins_cache();
+		$result = array( 'messages' => $skin->get_upgrade_messages() );
+		if ( $install_ok ) {
+			$new_plugins      = get_plugins();
+			$new_plugin_names = array_keys( $new_plugins );
+			if ( is_null( $plugin_name ) ) {
+				$diff_plugins = array_diff( $new_plugin_names, $old_plugin_names );
+				if ( empty( $diff_plugins ) ) {
+					$install_ok = false;
+				} else {
+					$plugin_name    = current( $diff_plugins );
+					$plugin_version = $new_plugins[ $plugin_name ]['Version'];
+				}
+			} else {
+				$plugin_info = $new_plugins[ $plugin_name ] ?? null;
+				if ( is_null( $plugin_info ) ) {
+					$install_ok = false;
+				} else {
+					$plugin_version = $plugin_info['Version'];
+				}
+			}
```

Comment:
I might be missing something here, but why can't we trust the result of `$upgrader->install()`? Why do we need to fetch the plugin name again this way / what would be the use case?

If we need to check something, we could maybe rely on `Plugin_Upgrader::plugin_info()` which should gives us the path to the file with plugin information. Similarly, there's some other information available in the (public) instance variable `$result` in `Plugin_Upgrader`.
Not sure if the diff approach is very solid.

Reply:
No, it's not you who is missing something, it's me: I just hadn't noticed the existence of the `plugin_info` method 🤦🏻‍♂️ So I have replaced all this code with a simple call to this method.